relay/pkg/block/forwarder.go

```diff
@@ -0,0 +1,110 @@
+package block
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-log"
+	"github.com/keep-network/tbtc/relay/pkg/btc"
+	"github.com/keep-network/tbtc/relay/pkg/chain"
+)
+
+const (
+	// Size of the headers queue.
+	headersQueueSize = 50
+
+	// Maximum size of processed headers batch.
+	headersBatchSize = 5
+
+	// Maximum time for which the pulling process will wait for a single header
+	// to be delivered by the headers queue.
+	headerTimeout = 1 * time.Second
+
+	// Block duration of a Bitcoin difficulty epoch.
+	difficultyEpochDuration = 2016
+
+	// Duration for which the forwarder should rest after performing
+	// a push action.
+	forwarderPushingSleepTime = 45 * time.Second
+)
+
+var logger = log.Logger("relay-block-forwarder")
+
+// Forwarder takes blocks from the Bitcoin chain and forwards them to the
+// given host chain.
+type Forwarder struct {
+	btcChain  btc.Handle
+	hostChain chain.Handle
+
+	processedHeaders int
+
+	headersQueue chan *btc.Header
+	errChan      chan error
+}
+
+// RunForwarder creates an instance of the block forwarder and runs its
+// processing loop. The lifecycle of the forwarder loop can be managed
+// using the passed context.
+func RunForwarder(
+	ctx context.Context,
+	btcChain btc.Handle,
+	hostChain chain.Handle,
+) *Forwarder {
+	forwarder := &Forwarder{
+		btcChain:     btcChain,
+		hostChain:    hostChain,
+		headersQueue: make(chan *btc.Header, headersQueueSize),
+		errChan:      make(chan error, 1),
+	}
+
+	go forwarder.pushingLoop(ctx)
+
+	return forwarder
+}
+
+func (f *Forwarder) pushingLoop(ctx context.Context) {
+	logger.Infof("running forwarder pushing loop")
+
+	for {
+		select {
+		case <-ctx.Done():
+			logger.Infof("forwarder context is done")
+			return
+		default:
+			logger.Infof("pulling new headers from queue")
+
+			headers := f.pullHeadersFromQueue(ctx)
+			if len(headers) == 0 {
+				continue
+			}
+
+			logger.Infof(
+				"pushing [%v] header(s) to host chain",
+				len(headers),
+			)
+
+			if err := f.pushHeadersToHostChain(ctx, headers); err != nil {
+				f.errChan <- fmt.Errorf("could not push headers: [%v]", err)
+				return
```

Comment:
Instead of pushing the error to the channel and hoping that the control loop will restart this process, why don't we just log an error and continue?

Reply:
Because we can enter into an inconsistent state. It's far better to fail here, restart the forwarder, initialize it with some chain data (it's needed by the pull side), and start from the beginning. This is also the way it was handled in the legacy relay maintainer but there the entire app crashed. Here we just restart the forwarding part.

Reply:
I have totally missed that forwarder is stateful. Can we leave a comment with an explanation? My proposition:
```
				f.errChan <- fmt.Errorf("could not push headers: [%v]", err)
				// We exit on the first error letting the code controlling the
				// relay to restart it. The relay is stateful and it is easier
				// to fetch the most recent information from BTC after the
				// restart instead of trying to recover here.
				return
```

Reply:
Sure. Please see: https://github.com/keep-network/tbtc/commit/71c059077a00c995b8eff208908ef2788873a92d